src/sinks/amqp.rs

```diff
@@ -0,0 +1,615 @@
+use crate::{
+    buffers::Acker,
+    config::{log_schema, DataType, GenerateConfig, SinkConfig, SinkContext, SinkDescription},
+    event::proto,
+    amqp::AmqpConfig,
+    sinks::util::{
+        encoding::{EncodingConfig, EncodingConfiguration},
+    },
+    template::{Template, TemplateParseError},
+    Event,
+};
+use futures::{
+    channel::oneshot::Canceled, future::BoxFuture, ready, stream::FuturesUnordered, FutureExt,
+    Sink, Stream,
+};
+use serde::{Deserialize, Serialize};
+use snafu::{ResultExt, Snafu};
+use std::{
+    collections::HashSet,
+    convert::TryFrom,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use prost::Message;
+use lapin::options::BasicPublishOptions;
+use lapin::BasicProperties;
+
+// Maximum number of futures blocked by [send_result](https://docs.rs/rdkafka/0.24.0/rdkafka/producer/future_producer/struct.FutureProducer.html#method.send_result)
+const SEND_RESULT_LIMIT: usize = 5;
+
+#[derive(Debug, Snafu)]
+enum BuildError {
+    #[snafu(display("creating kafka producer failed: {}", source))]
+    AmqpCreateFailed { source: lapin::Error },
+    #[snafu(display("invalid exchange template: {}", source))]
+    ExchangeTemplate { source: TemplateParseError },
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct AmqpSinkConfig {
+    pub(crate) exchange: String,
+    pub(crate) key_field: Option<String>,
+    pub(crate) encoding: EncodingConfig<Encoding>,
+    pub(crate) connection: AmqpConfig,
+}
+
+impl Default for AmqpSinkConfig {
+    fn default() -> Self {
+        Self {
+            exchange: "vector".to_string(),
+            key_field: None,
+            encoding: Encoding::Text.into(),
+            connection: AmqpConfig::default(),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, Derivative, Deserialize, Serialize, Eq, PartialEq)]
+#[serde(rename_all = "snake_case")]
+pub enum Encoding {
+    Text,
+    Json,
+    Protobuf,
+}
+
+pub struct AmqpSink {
+    _connection: lapin::Connection,
```

Comment:
Do we need to hold onto the connection too? I would have thought holding onto the channel would be sufficient.